src/extended_type_info_typeid.cpp

```diff
@@ -95,7 +95,7 @@ BOOST_SERIALIZATION_DECL void
 extended_type_info_typeid_0::type_unregister()
 {
     if(NULL != m_ti){
-        BOOST_ASSERT(! singleton<tkmap>::is_destroyed());
+        // Check that the singleton is still alive (might be unloaded already)
         if(! singleton<tkmap>::is_destroyed()){
```

Comment:
See #128:

- Assume 2 libraries using `extended_type_info_typeid_0` (singletons) with different types. Lets call these singletons `type<Foo>` and `type<Bar>`
- On initialization time both types register themselves in `singleton<tkmap>`, which hence gets constructed before.
  - Note that each library has its own instance of `singleton<tkmap>` due to static linkage of boost
- On destruction time library A gets unloaded: `type<Foo>` gets destructed, unregisters itself in `singleton<tkmap>`, afterwards `singleton<tkmap>` gets destructed.
- Due to those destruction issue in shared libs **both** `singleton<tkmap>` get destructed (you can check this easily with PRINTF in #128)
- Now when `type<Bar>` gets destroyed the `singleton<tkmap>` it wants to access is already destroyed
  - Previously this caused a BOOM -> See #128
  - Now the assertion would fire and the runtime check catches this.
  - No unregistering is required as the register does not exist altogether.
  - All good, just skip the unregistering code

Reply:
I expect that most singletons used in the serialization library to be constructed pre-main time and so in these cases I would expect is_destroyed to always return false.  BUT questions arise regarding this for dynamically loaded shared libraries.  Each shared library probably has it's own set of singletons.  Since shared libraries are not addressed by the standard and behavior is not documented anywhere that I know of, I don't know this for sure.  But it seems like the only reasonable way to do it.  In such a case I can see how a single might be created / destroyed before/after main is called.  Still since usage of such a singleton would be confined to the interior of the library, things should be OK.  Of course if the same type is used (maybe only declared in a header) things could get more complex - maybe two singletons of the same type might exist???.   Because of all this - and because I'm paranoid in general, I created is_destroyed basically to support runtime asserts and to gate things which might happen and might be legal.  It's basically defensive programming.  Down side is that it's really hard to guarantee that is_destroyed has the value that I expect of it.  In a simple case where types aren't dynamically loaded - I feel OK about this.  In other cases - it's much harder to say.  Given that all this depends so much on other/unknowable things, I don't think anyone can say much more than this.   We're stuck with dependence on testing.